### PR TITLE
Connection: move the notice for deleting connection owner

### PR DESCRIPTION
**Pull request by sergeymitr**
## Proposed changes:
* Move the "Delete Connection Owner" notice from JITM to Connection package.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
p9dueE-14B-p2#comment-1378
1191179647901802-as-1146698724415329

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
1. Connect Jetpack, create another user (non-connected).
2. Log in as the second user, log into WPCOM under different account.
3. Go to "Users", click "Delete" for the connection owner.
4. You should see the Jetpack notice. Compare it to the screenshot below. Screen Shot 2023-01-26 at 19 34 40
5. Click "Connect to Jetpack". Finish connection, reload the "Delete user" page.
6. The notice show now suggest you to replace the connection owner with another user. Compare it to the screenshot below. Screen Shot 2023-01-26 at 19 38 15
7. Choose a new connection owner, click "Set new connection owner". The notice should disappear, you can now delete the user.
8. Delete the user, confirm that you see no other notices.
9. Confirm connection ownership got transferred (you can use Jetpack Backup to see the connection owner ID). There's a know issue with the connection owner sometimes not switching fully, but it's being dealt with separately.

### CODE REVIEW

**Comment by fgiannar** [comment]
Haven't reviewed however it's worth mentioning here that this is the second PR that attempts to move this notice to the Connection package.

@sergeymitr Here's the [first one ](https://github.com/Automattic/jetpack/pull/17136) (still open) with some great feedback from @anomiex

**Comment by sergeymitr** [comment]
Thanks @fgiannar, some good feedback there.

Most of it is about the endpoint though, which you already moved to the Connection package in #20067.

There are a couple of notices on the AJAX request error reporting though.
My plan is to use a separate PR to rewrite this code in vanilla JS, as we're gradually getting rid of jQuery usage. I'll make sure to include error handling there.
At this point we move the JS code as is. We don't make the things worse, so I guess that's all right to leave it without error handling as long as I actually get back to it with the improvements soon after merging.
What do you think?

**Comment by sergeymitr** [comment]
Hi @anomiex.
Since you reviewed [the previous attempt](https://github.com/Automattic/jetpack/pull/17136) to move that code, can you please give this PR a look when you get a few minutes.

File: projects/js-packages/svelte-data-sync-client/changelog/update-move-delete-connection-owner-notice [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
```
Comment (anomiex) [inline projects/js-packages/svelte-data-sync-client/changelog/update-move-delete-connection-owner-notice]: Looks like the auto-format got done in a different PR brought in via trunk merge, so this could be deleted.

File: projects/packages/connection/src/class-connection-notice.php [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+					$( '#jp-switch-connection-owner' ).on( 'submit', function( e ) {
+						var formData = $( this ).serialize();
+						var submitBtn = document.getElementById( 'jp-switch-connection-owner-submit' );
+						var results = document.getElementById( 'jp-switch-user-results' );
+
+						submitBtn.disabled = true;
+
+						$.ajax( {
+							type        : "POST",
+							url         : "<?php echo esc_url( get_rest_url() . 'jetpack/v4/connection/owner' ); ?>",
```
Comment (anomiex) [inline projects/packages/connection/src/class-connection-notice.php:160]: This will probably work (no `"` or `\` or `<` should make it out, and no `&` or `'` should be in the input). But to be more correct, this should probably be
```suggestion
							url         : <?php echo wp_json_encode( esc_url_raw( get_rest_url() . 'jetpack/v4/connection/owner' ), JSON_HEX_TAG | JSON_HEX_AMP ); ?>,
```
The problem is that `esc_url` will entity-encode ampersands and single-quotes, but in JS we need to apply JS string encoding.

File: projects/packages/connection/src/class-connection-notice.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+							data        : formData,
+							headers     : {
+								'X-WP-Nonce': "<?php echo esc_js( wp_create_nonce( 'wp_rest' ) ); ?>",
```
Comment (anomiex) [inline projects/packages/connection/src/class-connection-notice.php:163]: As above, the nonce shouldn't contain anything that would make this problematic (`"` or `'` or `<` or `>` or `&`) but to really be correct it should be like
```suggestion
								'X-WP-Nonce': <?php echo wp_json_encode( wp_create_nonce( 'wp_rest' ), JSON_HEX_TAG | JSON_HEX_AMP ); ?>,
```
Again, the problem is that `esc_js` is intended for single-quoted strings embedded in tag inline attributes (like `onclick=""`) and so entity-encodes various characters.

File: projects/packages/connection/src/class-connection-notice.php [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+							},
+							success: function() {
+								results.innerHTML = "<?php esc_html_e( 'Success!', 'jetpack-connection' ); ?>";
```
Comment (anomiex) [inline projects/packages/connection/src/class-connection-notice.php:166]: As above, probably ok (here we _want_ the entity encoding, and it's unlikely anyone will put `\` in a translation) but to really be correct it should pass it through `wp_json_encode` too to ensure it's valid JS.
```suggestion
								results.innerHTML = <?php echo wp_json_encode( esc_html__( 'Success!', 'jetpack-connection' ), JSON_HEX_TAG | JSON_HEX_AMP ); ?>;
```

File: projects/packages/connection/src/class-connection-notice.php [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+								setTimeout( function() {
+									$( '#jetpack-notice-switch-connection-owner' ).hide( 'slow' );
+								}, 1000 );
+							}
+						} ).done( function() {
+							submitBtn.disabled = false;
+						} );
```
Comment (anomiex) [inline projects/packages/connection/src/class-connection-notice.php:173]: I see no error handling in the ajax call. Even the `.done()` callback here is equivalent to the `success` property above, so the button won't even be re-enabled.

I'd recommend bringing in e6540b854c4ae1ad9014cf6582dac6f8a14e1cea, which will at least output a "something went wrong" message if the call fails. Even better if you can give something more specific (does the API endpoint return anything useful that could be relayed?).

**Review by anomiex — COMMENTED** [review]
Just read through, didn't test it myself.

Pointed out some escaping stuff that should work but isn't technically correct. I won't insist if you want to stick with the "will work" code.

Also recommending bringing in the error handling that was added in the previous attempt.

File: projects/js-packages/svelte-data-sync-client/changelog/update-move-delete-connection-owner-notice (hunk 1 above)
Comment (sergeymitr) [reply projects/js-packages/svelte-data-sync-client/changelog/update-move-delete-connection-owner-notice]: Removed, thanks!

**Comment by sergeymitr** [comment]
Upon discussing the issue with @anomiex and @fgiannar, we decided to leave the existing JS code as is, and address the issues listed above in a separate PR along with refactoring that JS code entirely.

**Review by darssen — DISMISSED** [review]
As the idea "for now" is to have the existing code in the connection package. I just made sure to test prior and after and the flow worked flawlessly with the new branch enabled.

**Comment by sergeymitr** [comment]
@darssen can you please re-approve it, I had to merge it and the approval got reset.

**Review by darssen — APPROVED** [review]

File: projects/packages/connection/src/class-connection-notice.php (hunk 2 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-connection-notice.php:160]: Done: https://github.com/Automattic/jetpack/pull/29087/commits/2384298df4acd8642a0448130ff6a4ee32ef1768

File: projects/packages/connection/src/class-connection-notice.php (hunk 3 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-connection-notice.php:163]: Done: https://github.com/Automattic/jetpack/pull/29087/commits/0e9d586c96700daefa479f8c151a2b9eb1319eb8

File: projects/packages/connection/src/class-connection-notice.php (hunk 4 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-connection-notice.php:166]: Done: https://github.com/Automattic/jetpack/pull/29087/commits/b9e5d2764affd27fe12a3d7ed0d31f850a84f4a4

File: projects/packages/connection/src/class-connection-notice.php (hunk 5 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-connection-notice.php:173]: Done: https://github.com/Automattic/jetpack/pull/29087/commits/a7403347ea72a3e1dd11f066626e73fe862ede39
